### PR TITLE
[core] Remove import thread away from function table.

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
Now, import thread is only useful for run functions on all worker and it's only set in driver. This PR removes import thread and all lock related things.
We'll do run functions on all work when we start the worker.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by iycheng** [comment]
Hopefully, it's working. I'm very happy about removing 400 LOC.

File: src/ray/protobuf/pubsub.proto [hunk 1]
```diff
@@ -47,8 +47,6 @@ enum ChannelType {
   RAY_ERROR_INFO_CHANNEL = 8;
   /// A channel for logs from various Ray components.
   RAY_LOG_CHANNEL = 9;
-  /// A channel for keys to pickled python functions and actor classes.
```
Comment (mwtian) [inline src/ray/protobuf/pubsub.proto:50]: `reserve` the tag number for now, so it is easy for others to see the tag number has been deleted, and they can choose to re-use / not-use the value?

**Review by mwtian — APPROVED** [review]

File: src/ray/protobuf/pubsub.proto (hunk 1 above)
Comment (iycheng) [reply src/ray/protobuf/pubsub.proto:50]: Given that backward compatibility is not a critical issue right now, are you ok with just adjusting the labels?

Comment (mwtian) [reply src/ray/protobuf/pubsub.proto:50]: Yeah this is minor. Reducing tag numbers of other fields by 1 is fine too.

**Comment by iycheng** [comment]
It seems there is a bug that will cause the core worker hang for some cases. I'm still investigating that.

**Comment by iycheng** [comment]
Tracking the cancel bug here https://github.com/ray-project/ray/issues/24359

**Comment by iycheng** [comment]
After fixing the bug in task cancelation, there is another issue with this approach and I'm thinking about whether we need to deprecate the feature: run functions on all workers.

Before this PR, the code path is:
1. driver/worker push run_function_on_all_worker, driver execute the function
2. when the worker started, it'll execute the function in a separate thread.
3. when running in the middle, we still are able to push tasks to run.

This will cause some race conditions: workers may begin to accept tasks before importing finished.
3) actually is not used in ray core and it's a private API, although it's used in modin for importing something. (https://github.com/modin-project/modin/blob/685f1b4eb60c3dd6c9127bee93b6c138f890f700/modin/core/execution/ray/common/utils.py#L215-L219)

With this PR, the code path is:
1. driver push the run_function_on_all_worker
2. worker won't notify raylet that we should start until the functions are executed successfully
3. when running in the middle, we are not able to push tasks to run.

This works perfectly, but 2) will be blocked if it can't get the init function and cpp/java will fail when start python worker (x-lang needs this feature) because there code is not updated to set the key.

Solutions:
1. keep the current logic: we have to keep the function table
2. setup the key in cpp/java: 1) we don't have API for these two langs right now (20LOC); 2) it's wired since they are setting up something for python to run.
3. use code_search_path and don't deprecate run_function_on_all_worker

I think 3) works the best for us, it's simple and can meet all the requirement and no race condition. 
Any ideas @ericl @mwtian ???

**Comment by mwtian** [comment]
I don't fully understand the x-lang aspect, but if switching `code_search_path` can remove avoid calling `run_function_on_all_worker()` from python workers, that seems like an improvement.

**Pull request closed without merging**